### DISCUSSION TITLE
Update download tip about uninstalling to include deleting the stuff folder

### DISCUSSION THREAD

**Discussion by DarrenTAnims** [opentoonz/opentoonz#6462] [Ideas]
Om the download page is a useful hint:
"Please make sure to uninstall all the previous versions of OpenToonz before installing V1.5."

However, most users will only run the uninstaller and leave their OpenToonz Stuff folder behind, but this is where all of their previous settings are that presumably could clash with 1.5 settings. So we should probably tell them explicitly:

"Please make sure to uninstall all the previous versions of OpenToonz and delete your OpenToonz Stuff folder before installing V1.5.
On Windows the default location is: C:\ OpenToonz Stuff
On Mac the default location is: ..."

@shun-iwasawa, If this change is needed, it should probably be made pretty soon before too many users install it and run into problems.
Thanks.

**Comment by RodneyBaker** [comment]
If the recommendation is to delete the stuff directory, that note to uninstall will need to also contain a warning about the potential loss of project (scene) files and drawings as that is where most project data will be stored for the average user.

This is why I recommend to rename the stuff directory rather than delete it.
I tend to use a method that appends the current date so that I know roughly the date of the data so that when looking for a project/scene I have a greater chance of finding it.

In #3670 there is a proposal to include in the installer an option to 'archive' the stuff directory.
That proposal goes a little further than simply renaming the stuff directory as it also is designed to all the user to create a portable copy of the program being replaced also.  In a perfect world that archival process would have several options which the user could choose from but each would take care of the problems associated with maintaining the old stuff directory... with the possible exception of an option to explicitly keep the contents of the current stuff directory and have the new installation overwrite files.

But... all of this to say that I personally would NOT recommend deleting the stuff directory without first creating a backup copy.  If deleting, the potential for files to be missing thereafter is very high.

**Comment by DarrenTAnims** [comment]
I'd be happy for the warning to suggest renaming that directory. I spent over an hour yesterday trying to help someone, only to discover they upgraded from 1.4 and didn't delete/rename their stuff folder. This will happen over and over again for the next year unless we do something now. And as we've got a warning about uninstalling, which on its own isn't enough, it makes sense to me to amend it with instructions for getting a clean install.
Until we get the upgrade tool of course :-)

**Comment by RodneyBaker** [comment]
I've mentioned this before but a quick method I use regarding the stuff directory is to simply append the current date.
Example:  'C:\Opentoonz stuff'   becomes   'C:\Opentoonz stuff (20210415)'

In this way when drag/dropped into a holding or archiving directory the stuff folders collected there automatically sort by date (year month day).

Once thusly renamed (even if not moved at that time from the C:\ root directory) I can proceed with a new installationknowing that everything in the stuff directory has been placed there by the new installation.   Nothing else remains there.

If then needing to grab a project file or scene I was recently working on I just have to look for directories of that general timeframe wherever these dated stuff directories are maintained.   Optionally, but generally recommended for backup purposes,  these stuff directories can then be copied to external storage awaiting the inevitable day when they are needed.

An automated backup system could dispense with the renaming and simply copy the current stuff directory into the archiving location but that approach lacks the gain of having a truly empty stuff folder which is optimal for a number of situations.  For instance, at the moment of creating a new stuff folder that is the optimum time to use that pristine data in a portable copy of the current installation.  In very short order that stuff directory will change and will no longer be in that optimal state.

**Comment by konero** [comment]
When you run the installer there's an option [to overwrite default stuff files excluding user data](https://github.com/opentoonz/opentoonz_installer_windows/pull/6), why not use this?

**Comment by DarrenTAnims** [comment]
Does that remove any old settings and overwrite all new settings with the ones you need? If so, then that might be just fine. But the person I was helping had a problem, so whether they'd done that I don't know. Maybe that could be highlighted on the download page if so.

**Comment by konero** [comment]
It overwrites everything created by the installer by default, except for user profile settings and project data.

Yes, if a new option was added to a later release, the OT executable has a hard-coded default value it will write out into existing profile settings if the key is not found when you launch it. If an option has been removed, the key will linger in your profile but it would cause no problem.

By the way, this option is enabled by default isn't it? 🤔

**Comment by RodneyBaker** [comment]
@konero I assume you are addressing the initial issue @DarrenTAnims reported.
Just to make it clearer, my additional thoughts address the deeper idea of removing everything from the stuff directory for a number of cases to include some not under consideration here by Darren's report.  For instance, the stuff directory can grow to fairly large size with projects and scenes not related to a current production.  Archiving the files and creating a new stuff directory resolves that.

Relating more to Darren's report:
A consideration is that an installation may for a number of reasons not replace all stuff files desired to be replaced (ex:  files are read only or in use at the time of installation, multiple installations that conflict, etc.  Whereas relying on the installer to properly update files is desirable and will often work, starting from a point where there is no stuff directory goes even farther by guaranteeing that all files and settings will be the new default.

There are other reasons as well which I touch upon in #3670.
For instance, let's say we decide to update our stuff directory and leave the old data there... simply letting the installer overwrite the old data with new.   We have now lost that older state and can never use it again with older scenes and projects.  For this reason I find it best to preserve the old state before creating the new.  With this approach we can not only return to a scene/project but also to the exact same program and settings used for that scene/project.  We can then bring that older production forward into the new programming environment if that is desired.

The opportune time to do this is when moving to a new production. 
That also tends to be the time when program versions get updated (to avoid feature creep etc.)

**Comment by artisteacher** [comment]
The macOS installer doesn't have an option to overwrite stuff files.

macOS location: Applications/OpenToonz/OpenToonz_stuff

**Comment by shun-iwasawa** [comment]
I'm quite hesitant to ask users to remove / rename the stuff folder before installing V1.5 because users may misunderstand that they cannot bring their work from the previous versions to V1.5.

I believe that OT keeps backward compatibility with the conventional stuff folder of older versions (as long as they install with activating the `Overwrite all setting files in the Stuff folder except user's personal settings` option) .
@DarrenTAnims I still have no idea with your case why the user could launch the software by removing the old stuff folder... but I think in the most cases OT V1.5 can launch by just removing the binaries of the previous version before installing.

I investigated what exact file is preventing the launch, by comparing contents of `C:\Program Files\OpenToonz` between after clean install and install without unsinstall of V1.4 (say, a "dirty" install).

Here are files remained only in the "dirty" installed folder:
- `bearer/qnativewifibearer.dll`
- `srv/bearer/qnativewifibearer.dll`
- `opencv_world412.dll`
- `opengl32sw.dll`

And, once I removed `bearer/qnativewifibearer.dll` while leaving the others, the program started to launch normally.

`qnativewifibearer.dll` seems to work something related to Wifi ... And it had been [removed](https://codereview.qt-project.org/#/c/214029/) from Qt 5.10 due to a [bug](http://bugreports.qt.io/browse/QTBUG-40332) and had replaced by another component from Qt 5.14.

As a side note, OT seems to render .mov file with no problem even if  `srv/bearer/qnativewifibearer.dll` does exist.

**Comment by gab3d** [comment]
@shun-iwasawa 
I don't know exactly why, but the problem keeps coming to us periodically.

Today on the Spanish OT group at Facebook I had one user showing his screenshot of 1.5 having a missing Viewer, but what it really surprised me was to see the Basics room of the UI, when it shouldn't be there. 🤷🏻‍♂️

Here te screenshot:
![FB_IMG_1618542487414](https://user-images.githubusercontent.com/18104091/114965895-e95d3b00-9e47-11eb-81b2-3f429336f84c.jpg)

**Comment by shun-iwasawa** [comment]
> but what it really surprised me was to see the Basics room of the UI, when it shouldn't be there.

Currently the user's workspace settings belong to user's personal settings, so they won't be overwritten on installation.
You need to use `Reset to Default Rooms` command to synchronize the latest preset.

For now I have no idea why the viewer disappeared, though..

**Comment by gab3d** [comment]
@shun-iwasawa 
Thanks anyway for the info, will be useful for future cases. 💪

**Comment by DarrenTAnims** [comment]
So if a user selects the option, "Overwrite all setting files in the Stuff folder except user's personal settings", they still won't see the new layouts until they reset. Will they have any new commands in the menus? Because I think the menus are stored with each room too.
It'd just be nice to be able to have a clean install. And for most programs, uninstalling the previous version ensures you have a clean install. And users won't understand that uninstalling previous version doesn't give you a clean install.

**Comment by konero** [comment]
> So if a user selects the option, "Overwrite all setting files in the Stuff folder except user's personal settings", they still won't see the new layouts until they reset. Will they have any new commands in the menus? Because I think the menus are stored with each room too.

When you use the reset rooms command it deletes the user room set folder entirely, then on next launch copies the default set into the user profile just like it does on first launch after an install, so it will grab new menu commands too as long as the overwrite option was ticked when updated.

> It'd just be nice to be able to have a clean install. And for most programs, uninstalling the previous version ensures you have a clean install. And users won't understand that uninstalling previous version doesn't give you a clean install.

Hmm I don't know about that, Clip Studio (and many others) will store their customizations data usually (on Windows) in `%USERPROFILE%/AppData/Roaming` as standard, so if I install a new version of Krita, it will be just as it was before with all my workspace customizations.

Sometimes when you run an uninstaller it will ask if you'd like to remove all user data, but that's usually only necessary if you no longer wish to use the program at all and just want to clean up left over files, if you're updating and intend to continue using then you almost certainly want to keep customizations I would imagine. If not, then the reset to default option will do the trick, e.g. in Clip Studio you do Window > Workspace > Reset to default.

**Comment by manongjohn** [comment]
The whole reason for keeping the stuff folder during upgrade is so you don't have to reset all your preferences and room layouts up again since those don't get overwritten. Also anything saved in sandbox of the stuff project folder and histories are maintained and you can continue working without spending time trying to reconfigure how you used to have it. Likely you will miss something.

The unfortunate side effect of that is rooms and menus don't update. This is good for those who customize their rooms, but bad for those that don't since they wont see new rooms and menu items.

Personally, I would rather deal with rooms not being updated than have to reset all my preferences and lose files and file history.

Perhaps if the installer detects an existing stuff folder, it should question the user about resetting room layouts or leave it as is.

**Comment by gab3d** [comment]
@manongjohn 
I agree with that proposal.
An option to "Reset Rooms and Menus to Default" at installation time would be sufficient.
I'd mark it by default, though.

**Comment by shun-iwasawa** [comment]
I tried making installer to detect the old version and uninstall it in advance.

Regarding resetting the room layouts and menus, I think it would be better to suggest resetting them not on installing but on runtime of OT, considering that multiple users can share the same stuff folder and there may be the case that some users would like to reset the settings and the others don't.

**Comment by DarrenTAnims** [comment]
99% of the users I speak with don't have multiple users on their computer, so resetting for all users isn't an issue. However, I understand the point. But if the page in the installer that asks about overwriting your settings also asked about resetting rooms, menus, etc, that could be a good intermediate fix.
I don't think we should expect the average user to step through the installer and then have to do something else after installing to see new options. I think the installer should ask all of the questions it needs, do all of the updating it needs and then on running, the user should be able to see all of the new features without "having to know" to click something else.

**Comment by RodneyBaker** [comment]
We need to review the entire downloading and installation process moving forward to v1.7... v2.0 and beyond.
I think the core of this report's request has been addressed by https://github.com/opentoonz/opentoonz_installer_windows/pull/22 but until we conduct a round or two of review and feedback I don't suppose we can know.

The one element of this request that I do want to avoid is any suggestion that the user DELETE their stuff directory.
That will surely lead to loss of files the user will not want to lose.

@DarrenTAnims I'll recommend reviewing your original intent versus what the current process provides and if the primary intent is met close this report.  More certainly can be done and I recommend additional reports be opened for other specific requirements related to download, installation and upgrade/updating processes.

Also relates to:  #3892

**Comment by artisteacher** [comment]
> I think the core of this report's request has been addressed by https://github.com/opentoonz/opentoonz_installer_windows/pull/22 but until we conduct a round or two of review and feedback I don't suppose we can know.

That doesn't really help those of us on macOS or Linux.

**Comment by RodneyBaker** [comment]
> That doesn't really help those of us on macOS or Linux.

A never ending battle for sure.
Follow up will always be needed to extend any solution to be fully cross platform.**
If the fix is in however for any of the three platforms then at least we can know there is a resolution and can identify that for porting to the other platforms.

**I suppose one question to ask related to that particular issue is:  Is there any fully cross platform install system we can adopt that would eliminate the need for different installers on each platform?  If not then how can we best align the current installers to maximize cross platform compatibility?

**Comment by RodneyBaker** [comment]
Recommend a review of current installers to determine what (if anything) should be adjusted moving into the v1.8 release.

Regarding this request to add a note to download page, it appears that will not be implemented.
Some options:   

- Shift focus of this report to other area (editing the report title according)
- Refine and restate the problem
- Close the report / open a new report as needed
- Something else...
